### PR TITLE
Distributed tracing support with Micrometer and OpenTelemetry

### PR DESCRIPTION
**Pull request by o-shevchenko**
Context: https://github.com/LogNet/grpc-spring-boot-starter/issues/362

### CODE REVIEW

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingClientInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.TraceId;
+
+public class TracingClientInterceptor implements ClientInterceptor {
```
Comment (o-shevchenko) [inline grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingClientInterceptor.java:13]: Should it be moved to grpc-client-spring-boot-starter?

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/log/LogInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class LogInterceptor implements ServerInterceptor {
```
Comment (o-shevchenko) [inline grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/log/LogInterceptor.java:12]: There is an example but LogInterceptor was not a part of grpc-spring-boot-starter. I can add it in a separate PR if needed

**Comment by o-shevchenko** [comment]
@jvmlet Any thoughts?

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingClientInterceptor.java (hunk 1 above)
Comment (jvmlet) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingClientInterceptor.java:13]: Yes

**Comment by jvmlet** [comment]
> @jvmlet Any thoughts?

Sorry, busy days. Will review the PR soon

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/autoconfigure/tracing/OtelConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
+import org.lognet.springboot.grpc.tracing.TracingServerInterceptor;
+import org.springframework.boot.actuate.autoconfigure.tracing.otlp.OtlpProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableConfigurationProperties(OtlpProperties.class)
+public class OtelConfiguration {
```
Comment (jvmlet) [inline grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/autoconfigure/tracing/OtelConfiguration.java:17]: this should be conditional on `OpenTelemetrySdk` class

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/log/LogInterceptor.java (hunk 2 above)
Comment (jvmlet) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/log/LogInterceptor.java:12]: Why  is this class needed ?

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.TraceFlags;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+
+public class TracingServerInterceptor implements ServerInterceptor {
```
Comment (jvmlet) [inline grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java:16]: Should have the highest  precedence https://github.com/LogNet/grpc-spring-boot-starter#421-interceptors-ordering

File: grpc-spring-boot-starter/build.gradle [hunk 5]
```diff
@@ -286,6 +286,8 @@ dependencies {
     compileOnly "org.springframework.boot:spring-boot-starter-actuator"
     compileOnly "org.springframework.boot:spring-boot-starter-validation"
     compileOnly 'org.springframework.cloud:spring-cloud-starter-consul-discovery'
+    compileOnly 'io.opentelemetry:opentelemetry-exporter-otlp:1.29.0'
+    compileOnly 'io.micrometer:micrometer-tracing-bridge-otel:1.1.4'
```
Comment (jvmlet) [inline grpc-spring-boot-starter/build.gradle:290]: `io.micrometer:micrometer-tracing-bridge-otel` is it in-use ?

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingClientInterceptor.java (hunk 1 above)
Comment (o-shevchenko) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingClientInterceptor.java]: done

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java (hunk 4 above)
Comment (o-shevchenko) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java:16]: Do you mean we need to add such info to the doc?
Like `Use @Order(10) with TracingServerInterceptor `?

File: grpc-spring-boot-starter/build.gradle (hunk 5 above)
Comment (o-shevchenko) [reply grpc-spring-boot-starter/build.gradle:290]: image
Used in OtelConfiguration, removed from client

File: grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java (hunk 4 above)
Comment (o-shevchenko) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java:18]: added @Order(20)? I don't see you use `@Order`. Have you had in mind any other mechanisms?

Comment (jvmlet) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java:18]: see [this](https://github.com/LogNet/grpc-spring-boot-starter#421-interceptors-ordering:~:text=and%20Metrics%20integration.-,Their%20order%20can%20also%20be%20controlled%20by%20below%20properties%20%3A,-grpc.recovery.interceptor), we will need to move `recovery` and  `auth` down `1`

Comment (o-shevchenko) [reply grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/tracing/TracingServerInterceptor.java]: I've set HIGHEST_PRECEDENCE. Do I need to touch recovery and auth here? Please, let me know if you want to change order somehow

**Comment by kprasad99** [comment]
probably better to use `grpc-instrumentation` dependency and set context, the above changes doesn't seems to work, header here considered is `traceId` and not `traceparent`, propagators can be different. Following are the steps I followed.

1. Add gradle dependency.
```
implementation "io.opentelemetry.instrumentation:opentelemetry-grpc-1.6:${otelInstrumentationVersion}"
```
3. Add Grpc otel configuration class. the below @Autoconfiguration is simpler one, however we might need to add additional conditionals to enable the tracer.
```
@Slf4j
@Configuration
@ConditionalOnEnabledTracing
public class GrpcOtelConfiguration {

   @Bean
    @GRpcGlobalInterceptor
    public ServerInterceptor otelGrpcInterceptor(OpenTelemetry telemetry) {
		log.info("Enabling otel tracer for grpc");
		return GrpcTelemetry.create(telemetry).newServerInterceptor();
	}
	
}

```

**Comment by pieter-stek** [comment]
Micrometer already supports [gRPC instrumentation](https://micrometer.io/docs/observation#_existing_instrumentations). So, using ObservationGrpcServerInterceptor and ObservationGrpcClientInterceptor from Micrometer is also an option

**Comment by o-shevchenko** [comment]
> Micrometer already supports [gRPC instrumentation](https://micrometer.io/docs/observation#_existing_instrumentations). So, using ObservationGrpcServerInterceptor and ObservationGrpcClientInterceptor from Micrometer is also an option

That's what I tried from the beginning. But it leads to the error:
```
Caused by: java.lang.RuntimeException: Failed to start GRPC server
	at org.lognet.springboot.grpc.GRpcServerRunner.start(GRpcServerRunner.java:112)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:284)
	... 18 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Failed to create interceptor instance.
	at org.lognet.springboot.grpc.GRpcServerRunner.lambda$bindInterceptors$3(GRpcServerRunner.java:129)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.lognet.springboot.grpc.GRpcServerRunner.bindInterceptors(GRpcServerRunner.java:140)
	at org.lognet.springboot.grpc.GRpcServerRunner.lambda$start$0(GRpcServerRunner.java:72)
	at java.base/java.util.HashMap.forEach(HashMap.java:1421)
	at org.lognet.springboot.grpc.GRpcServerRunner.start(GRpcServerRunner.java:66)
	... 19 common frames omitted
Caused by: java.lang.InstantiationException: io.micrometer.core.instrument.binder.grpc.ObservationGrpcServerInterceptor
	at java.base/java.lang.Class.newInstance(Class.java:639)
	at org.lognet.springboot.grpc.GRpcServerRunner.lambda$bindInterceptors$3(GRpcServerRunner.java:127)
	... 34 common frames omitted
Caused by: java.lang.NoSuchMethodException: io.micrometer.core.instrument.binder.grpc.ObservationGrpcServerInterceptor.<init>()
	at java.base/java.lang.Class.getConstructor0(Class.java:3585)
	at java.base/java.lang.Class.newInstance(Class.java:626)
	... 35 common frames omitted
```

**Comment by o-shevchenko** [comment]
The issue can be avoided if we create a bean manually:
```
@Bean
    fun serverInterceptor(observationRegistry: ObservationRegistry): ObservationGrpcServerInterceptor {
        return ObservationGrpcServerInterceptor(observationRegistry);
    }
    ``

**Comment by o-shevchenko** [comment]
ObservationGrpcServerInterceptor doesn't provide integration with SpanContext. It doesn't add headers into context, so I can't log traceId and spanId

**Comment by pieter-stek** [comment]
> ObservationGrpcServerInterceptor doesn't provide integration with SpanContext. It doesn't add headers into opentelemetry Context, so I can't log traceId and spanId

Spring Boot supports both OpenTelemetry and OpenZipkin Brave for tracing. Micrometer is used as an abstraction layer over both libraries. I think grpc-spring-boot-starter should also support both tracing libraries, so only Micrometer API's should be used and no OpenTelemetry API's.

**Comment by o-shevchenko** [comment]
yes, but I don't see in micrometer ability to send traces to gRPC collector `otlp.tracing.endpoint: http://collector:4317`

**Comment by o-shevchenko** [comment]
Right, that's what I currently use. But I can't get traceId and spanId in logback
I used `TracingServerInterceptor ` to put my headers into Context manually, so it's available in MDC.

```
logging.pattern.level: "trace_id=%mdc{trace_id} trace_id=%mdc{x-b3-traceid} trace_id=%X{trace_id} trace_id=%X{traceId:-} span_id=%mdc{spanId} %p"
```

**Comment by o-shevchenko** [comment]
No, I've tried everything. GrpcServerObservationContext is not propagated to GrpcServerObservationContext in GRpcService

**Comment by vicmosin** [comment]
@o-shevchenko did you find any way after all?

**Pull request closed without merging**